packages/fpdart/lib/src/effect.dart

```diff
@@ -0,0 +1,800 @@
+              (exit) => signal
+                  .failCause<E, L>(const Interrupted())
+                  .__unsafeRun(context.withoutSignal)
+                  .then((_) => exit),
+            );
+      });
+
+  /// {@category constructors}
+  static Effect<E, L, void> sleep<E, L>(Duration duration) => Effect.from(
+        (_) => Future.delayed(
+          duration,
+          () => const Right(null),
+        ),
+      );
+
+  /// {@category constructors}
+  static Effect<E, Never, Never> die<E>(dynamic defect) => Effect.from(
+        (_) => Left(Die.current(defect)),
+      );
+
+  /// {@category constructors}
+  static Effect<E, Never, Never> functionDie<E>(dynamic Function() run) =>
+      Effect.from(
+        (_) => Left(Die.current(run())),
+      );
+
+  /// {@category constructors}
+  static Effect<E, L, fpdart_unit.Unit> unit<E, L>() => Effect.from(
+        (_) => const Right(fpdart_unit.unit),
+      );
+
+  /// {@category collecting}
+  static Effect<E, L, Iterable<R>> forEach<E, L, R, A>(
+    Iterable<A> iterable,
+    Effect<E, L, R> Function(A a, int index) f,
+  ) =>
+      Effect.from(
+        (context) {
+          if (iterable.isEmpty) {
+            return const Right([]);
+          }
+
+          return iterable
+              .mapWithIndex(f)
+              .fold<Effect<E, L, Iterable<R>>>(
+                Effect.succeed(const Iterable.empty()),
+                (acc, effect) => acc.zipWith(
+                  effect,
+                  (list, r) => list.append(r),
+                ),
+              )
+              ._unsafeRun(context);
+        },
+      );
+
+  /// {@category collecting}
+  static Effect<E, L, Iterable<R>> all<E, L, R>(
+    Iterable<Effect<E, L, R>> iterable,
+  ) =>
+      Effect.forEach(iterable, (a, _) => a);
+
+  /// {@category zipping}
+  Effect<E, L, C> zipWith<B, C>(
+    Effect<E, L, B> effect,
+    C Function(R r, B b) f,
+  ) =>
+      flatMap(
+        (r) => effect.map(
+          (b) => f(r, b),
+        ),
+      );
+
+  /// {@category zipping}
+  Effect<E, L, R> zipLeft<C>(
+    Effect<E, L, C> effect,
+  ) =>
+      flatMap(
+        (r) => effect.map(
+          (_) => r,
+        ),
+      );
+
+  /// {@category zipping}
+  Effect<E, L, C> zipRight<C>(
+    Effect<E, L, C> effect,
+  ) =>
+      flatMap((_) => effect);
+
+  /// {@category context}
+  Effect<V, L, R> mapContext<V>(Context<E> Function(Context<V> context) f) =>
+      Effect.from(
+        (context) => _unsafeRun(f(context)),
+      );
+
+  /// {@category context}
+  Effect<V, L, R> mapEnv<V>(E Function(V env) f) => Effect.from(
+        (context) => _unsafeRun(
+          Context(env: f(context.env), signal: context.signal),
+        ),
+      );
+
+  /// {@category context}
+  Effect<Null, L, R> provide(Context<E> context) {
+    final env = context.env;
+    final effect = env is ScopeMixin && !env.scopeClosable
+        ? alwaysIgnore(env.closeScope())
```

Comment:
Hmm this could be an issue, as if there was a defect when try to close the scope, it will be swallowed / unhandled.

Reply:
You need to close the scope regardless of success or failure.